Add entity script for balloon spawner and sample json entity with us…

rabelaiis opened the pull request:
…erData to attach script to

hifi-gustavo commented:
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

rabelaiis commented on scripts/tutorials/entity_scripts/balloonSpawner/spawnBalloons.js:
```diff
@@ -0,0 +1,181 @@
+    var YELLOW_BALLOON_URL = "https://s3.amazonaws.com/hifi-rabelaiis/yellowBalloon.fbx";
+    var ORANGE_BALLOON_URL = "https://s3.amazonaws.com/hifi-rabelaiis/orangeBalloon.fbx";
+    var CYAN_BALLOON_URL = "https://s3.amazonaws.com/hifi-rabelaiis/cyanBalloon.fbx";
+    var PURPLE_BALLOON_URL = "https://s3.amazonaws.com/hifi-rabelaiis/purpleBalloon.fbx";
```
I think these models should be moved to the hifidelity servers and the urls updated

rabelaiis commented on scripts/tutorials/entity_scripts/balloonSpawner/balloonSpawner.json:
```diff
@@ -0,0 +1,20 @@
+                "z": 2
+            },
+            "name": "Balloon Spawner",
+            "serverScripts": "https://s3.amazonaws.com/hifi-rabelaiis/spawnBalloons.js",
```
This url should also needs to be updated. I'm not sure where it will end up on the hifi server

rabelaiis commented:
## Test plan
Note that this should probably be tested after the urls have been switched to point at high fidelity's servers to make sure everything is hooked up correctly. 

Load a balloon spawner using balloonSpawner.json by going to the create (used to be edit) menu and clicking import entities(.JSON). You should see an entity with a countdown that spawns balloons with random colors and random x, z starting positions. The balloons should float up and then be deleted after 10 seconds.

Modify the userdata for the balloon spawner under properties in the create menu. Verify that modifications to the userData changes how the balloon spawner works. There are six userdata properties to verify.
spawnRate - how fast balloons are spawned in ms
xRangeMax and zRangeMax - how far out balloons are randomly spawned
gravityCoefficient - how fast balloons float up
balloonLifetime - how long until the balloons are deleted in seconds
spawnTime - how long the spawner will spawn balloons for in seconds

rabelaiis commented:
The urls are now pointing to the hifi servers so this is ready for QA now

Nex-Pro commented:
Nice spawner.
You might want to guard against users putting low or negative numbers in spawnRate

Low numbers break it one way and negative values stop it working entirely.

When the spawnTime has expired there is simply an invisible primitive left behind. Is this intended ?

Also how about some user ( not programmer ) friendly names for the values. 

Because it is too easy to break I'll let someone else decide if this is a pass.

rabelaiis commented:
Thanks for the thorough testing. 
I added a guard for the spawnRate so that it doesn't go below 10 and prints a message if the user tries to set it below 10. Things on my server slowed down at 10 but nothing crashed. So it seemed like an ok limit if someone wants to push it.

I wasn't sure about keeping the parent primitive around but after talking to Jazmin about what she intends to use the spawner for, I think it makes sense to clean it up so I made that change as well.
I changed the 'spawnTime' name to be 'spawnDuration'. The marketplace description has a more detailed explanation of each parameter which should clear things up if there is any confusion on the user data names.

ZappoMan commented on scripts/tutorials/entity_scripts/balloonSpawner/spawnBalloons.js:
```diff
@@ -0,0 +1,222 @@
+        var parentProperties = Entities.getEntityProperties(pEntityID, ["userData"]),
+            spawnMusicURL,
+            spawnerSettings;
+        if (parentProperties.userData){
```
space between `)` and `{`

ZappoMan commented on scripts/tutorials/entity_scripts/balloonSpawner/spawnBalloons.js:
```diff
@@ -0,0 +1,222 @@
+            Entities.addEntity(balloonProperties);
+
+            //Clean up after spawnDuration
+            if (spawnCount * spawnRate / 1000 > spawnDuration){
```
coding standard: no magic numbers

ZappoMan commented on scripts/tutorials/entity_scripts/balloonSpawner/spawnBalloons.js:
```diff
@@ -0,0 +1,222 @@
+        spawnRate =  !isNaN(spawnerSettings.spawnRate) ? spawnerSettings.spawnRate : spawnRate;
+        spawnMusicVolume = !isNaN(spawnerSettings.spawnMusicVolume) ? spawnerSettings.spawnMusicVolume : 0.1;
+
+        if (spawnRate < 10){
```
coding standard: no magic numbers

ZappoMan replied:
space between ) and {

ZappoMan commented on scripts/tutorials/entity_scripts/balloonSpawner/spawnBalloons.js:
```diff
@@ -0,0 +1,222 @@
+            musicInjector.stop();
+            musicInjector = undefined;
+        }
+        if (pEntityID){
```
space between ) and {

ZappoMan commented:
@rabelaiis - can you make the updates to the code to match the coding standard? Thanks

rabelaiis commented:
@ZappoMan yes, I'll make the changes tonight. I've been travelling without a computer for the past few days.

rabelaiis commented:
@ZappoMan coding standard issues have been fixed. Thanks for the review!

marko8904 commented:
add to whitelist

hifi-gustavo commented:
Build console output for commit: 7f8f870d9f52bdfc7d96eb06cd0e45260a865302 available at: https://gist.github.com/6ea209f93cdbb681f112791df31a74a2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10372-7f8f870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10372-7f8f870.dmg

hifi-gustavo commented:
Build console output for commit: 7f8f870d9f52bdfc7d96eb06cd0e45260a865302 available at: https://gist.github.com/6ea209f93cdbb681f112791df31a74a2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10372-7f8f870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10372-7f8f870.dmg

Nex-Pro commented:
@ZappoMan Should I jump on this again ?

cozza13 commented:
@Nex-Pro yes please test

Nex-Pro commented:
ok

Nex-Pro commented:
### FAIL.

Music will not play on any  build I've tried.  

Checked the WAV file and it plays ok in VLC media playa when downloaded from the link.

rabelaiis commented:
That's odd. I'll take a look tonight.

rabelaiis commented:
As far as I can tell. The music not playing is something related to the build and not the balloon party code. The looping sound emitter in the marketplace also isn't working on the builds provided by gustavo. I merged in master again which might fix the problem. 

Since the balloon party code is all scripts, it shouldn't depend on the build so you should be able to test it on the current beta 6561 (https://highfidelity.io/download) which plays the sound correctly when I try it there.

hifi-gustavo commented:
Build console output for commit: d36593161159a0695342bfda279628a3f8aa78dc available at: https://gist.github.com/6ea209f93cdbb681f112791df31a74a2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10372-d365931.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10372-d365931.dmg

rabelaiis commented:
I think the issue with the audio was fixed with https://github.com/highfidelity/hifi/pull/10495.
You'll notice that no audio works not just the balloon spawner. They merged the fix into master this morning: https://github.com/highfidelity/hifi/pull/10507. I just re-merged master into my branch. The audio should be working again once the build finishes.

Nex-Pro commented:
Cool

Nex-Pro commented:
Retesting

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/42/HighFidelity-Beta-PR10372-5e8a844f789c6d93689a8dc9cd8ff7d1a345b82f.exe

marko8904 commented:
Mac OS X terminated connection during build. Rebuilding.

Nex-Pro commented:
Countdown takes place Balloons launch music plays.
Changing the user data works.
However. While this script probably meets the design spec. 
It is too easy to make a server grind to a halt with it as it stands.
I set the balloon lifetime to -1 and spawnrate( bad name )  at 10. Within a few seconds the client became unresponsive. 
This then leaves the difficult/boring task of removing the balloons.

Nex-Pro commented:
@marko it's a script test.

marko8904 commented:
test this please

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/48/HighFidelity-Beta-PR10372-5e8a844f789c6d93689a8dc9cd8ff7d1a345b82f.exe

rabelaiis commented:
I think the provided defaults are reasonable. I'd rather not restrict balloon lifetime functionality to protect against extreme inputs.

Nex-Pro commented:
I disagree.

Nex-Pro commented:
IMO this is unfit for release as is.

rabelaiis commented:
@Nex-Pro How would you recommend restricting the user inputs further?

Nex-Pro commented:
10 is an allowed input for spawn rate as is -1 lifetime result  25 minutes spent cleaning the server up

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/48/HighFidelity-Beta-PR10372-5e8a844f789c6d93689a8dc9cd8ff7d1a345b82f.dmg

Nex-Pro commented:
Set an upper limit on balloon life and also exclude -1 which is permanent.
I'd push all the new balloon ids into an array and set a limit on the allowable number of balloons in it.
These are not things you have to act on just on my recommendation. You should discuss them with the person that put this up on worklist.  There's a limit on how many physical/collidable entities can be present near any avatar. It is too easy with this to totally bork the server.

Nex-Pro commented:
@rabelaiis I had to write a script just to clean the server up.

rabelaiis commented:
@cozza13 @themelissabrown let me know if you want to add additional limitations on how long the balloon lifetime is and/or how many balloons you can spawn. 
I'm not sure what 'good' limits would be as different servers will be able to handle different amounts of entities. So I'm in favor of leaving it up to the user to determine what their server can handle. I also think the user may want to keep the balloons that they spawn around if they select that in the userdata.

themelissabrown commented:
@rabelaiis apologies. I was out of town. Will confirm an answer for you and get back to you ASAP

themelissabrown commented:
@rabelaiis We agree that this could be implemented via userdata controls: if the lifetime on the balloons is set to -1 so they are permanent, the balloons should be added to an array that gets cleaned up when the script stops running (spawner entity is deleted)
You can still filter on "balloon"
In the entity list
This could also be added to the user data
E.g. "ShouldCleanUpBalloonsOnEnding" so that users can toggle whether or not to keep the balloons once they have been spawned.

rabelaiis commented:
Those implementations make sense to me @themelissabrown. What are the next steps here? Should we set up another worklist job? I'm going on vacation tomorrow and wouldn't be available to work on it until June 5th.

Nex-Pro commented:
@themelissabrown @rabelaiis Thanks.

themelissabrown commented:
@rabelaiis hope you had a great trip! Feel free to add commits here and I'll get a WL job set up for the additional work.

rabelaiis commented:
@themelissabrown I took another look at the suggested changes and wanted to note that they duplicate the balloonlifetime field that is already available in user data.
To outline the two situations:

1. User wants to keep the balloons
Currently: Sets balloon lifetime to -1
Suggested changes: Sets balloon lifetime to -1 and sets 'ShouldCleanUpBalloonsOnEnding' to false

2. User wants the balloons cleaned up
Currently: Sets the balloon lifetime to be the spawner lifetime
Suggested changes: Sets the balloon lifetime to be -1 and sets 'ShouldCleanUpBallonsOnEnding' to true

The changes provide the same functionality and they don't seem clearer to me albeit I wrote the code. You can still end up in situations where lots of balloons are spawned that are not cleaned up. If you think the changes make the cleanup process easier to understand for the user or see other advantages and would still like me to proceed with the changes through the new WL job, let me know.

themelissabrown commented:
@rabelaiis https://worklist.net/21388

rabelaiis commented:
Thanks @themelissabrown, I saw the worklist job. Do you still want me to proceed with the changes even though they are redundant with the current functionality? See my comment on this pull request from yesterday for more context: https://github.com/highfidelity/hifi/pull/10372#issuecomment-306630397

themelissabrown commented:
@rabelaiis the default should be that the balloons clean themselves up and that the user can then change those settings if they want. A lot of our users won't understand what they have to do to manually change the settings and we don't want them frustrated, so the second option should be the default.

rabelaiis commented:
@themelissabrown that is already the default. I set it up so new users would have everything cleaned up automatically. It's only when the user goes into the settings and changes the balloon cleanup setting that the balloons are not cleaned up. I'll try and pull up with you on Skype today or tomorrow to make sure we are on the same page so we can close out this issue.

themelissabrown commented:
I think we can set this as done and ready to merge for now. If the balloons cause issues with frame rate or anything, we can have you add in code to put a hard upper limit on the lifetime of the balloons and on the qty spawned per second (or other unit of time that makes sense). Thank you!

rabelaiis commented:
@themelissabrown great thanks!
